### PR TITLE
feat(core): extend smoke gate with goodExample over-matching check

### LINKED ISSUE #1580: core: extend smoke gate to verify goodExample does not match (closes over-matching defect class)
## Context

The compile-time smoke gate (#1415) currently validates that a rule fires on its \`badExample\` (under-matching check). It does not validate that the rule does NOT fire on its \`goodExample\`. Over-matching is the dominant defect class observed in the 2026-04-18 security-pack postmerge incident (10-of-10 bad rate from #1526).

Closes #1434.

## Scope

- Flip \`CompilerOutputSchema.goodExample\` from optional to required for both ast-grep and regex engines. (Already required for \`badExample\` per #1420; this makes the pair symmetric.)
- Extend the smoke gate to run the compiled rule against \`goodExample\` and assert zero matches.
- If \`goodExample\` matches, reject the rule as \`nonCompilable\` with a new reason code (e.g., \`over-matching\` or \`matches-good-example\`).
- Tests: a rule that fires on goodExample gets rejected at compile time and surfaces in the nonCompilable ledger with the correct reason code.

## Gating

Blocks 1.15.0 external ship. The Zero-Trust thesis (ADR-089) requires that rule generation be bracketed by deterministic gates on both sides. Asymmetric verification (badExample only) is the biggest ergonomic leak.

## Related

- #1434 — original filing (this ticket is the execution)
- #1420 — CompilerOutputSchema.badExample required flip precedent
- #1526 — security-pack postmerge incident (10-of-10 bad rate)
- #1523 — compile-worker meta-lesson classifier (orthogonal, upstream gate)

### PR DESCRIPTION
**Pull request by satur8d**
## Summary

- `CompilerOutputSchema.goodExample` flips from optional to engine-conditional required (regex + ast-grep), mirroring the #1420 pattern for `badExample`. The `ast` engine stays exempt.
- `CompiledRuleSchema.goodExample` stays optional on the persisted-rule boundary for backward compat with pre-#1580 rules.
- `runSmokeGate` parameter renamed from `badExample` to `snippet` since it now serves both roles; docstring updated.
- `buildCompiledRule` adds the parallel over-matching check after the existing under-matching check: rule must fire on `badExample` AND must NOT fire on `goodExample`.
- Two new `NonCompilableReasonCodeSchema` values: `matches-good-example` (over-match rejection) and `missing-goodexample` (defensive path when callers bypass the schema refine).
- Pipeline 3 threads `snippets.good.join('\n')` as `goodExampleOverride`; Pipeline 2 requires LLM emission via the updated compiler prompt. Pipeline 1 (manual) is unaffected — gate stays opt-in via `enforceSmokeGate`.
- `extractGoodExample` helper added in `lesson-pattern.ts` (Pipeline 1 heading-based extraction, parallel to `extractBadExample`).

## Why

Over-matching is the dominant defect class observed in the 2026-04-18 security-pack postmerge incident (10-of-10 bad rate from #1526). The existing smoke gate caught under-matching; nothing structural caught a rule that fires on both its badExample AND every legitimate call site. This ticket closes that half.

## Shield self-review round

First commit (`95cfd5e2`) drew two CRITICAL findings from Shield:

1. Missing tests for `extractGoodExample` (and by extension `extractBadExample`, which had been under-tested since #1408). Added 14 tests in `lesson-pattern.test.ts` covering fence styles, heading casing, empty blocks, and bad+good sibling extraction.
2. Empty-string `goodExampleOverride` trap: when a Pipeline 3 lesson's Good block is empty, `snippets.good.join('\n')` is `''`, and `options.goodExampleOverride ?? parsed.goodExample` clobbers the LLM-echoed value because nullish coalescing treats `''` as defined. Pipeline 3 call site now passes `undefined` when `snippets.good.length === 0`; symmetric guard added for bad. Two tests pin the fallback contract and the trap it protects against.

Both fixed in `208816bb`. Shield second-round review PASS with no findings.

## Test plan

- [x] 1249 core tests pass (16 net-new: 5 smoke-gate + 8 build-compiled-rule integration + 14 lesson-pattern extraction — minus overlap on shared fixtures)
- [x] Full workspace `pnpm test` green (8/8 suites)
- [x] `totem lint` PASS
- [x] `totem review` PASS (second round)
- [x] `totem verify-manifest` PASS (440 rules)
- [x] Changeset lands a patch bump across the fixed group

## Follow-ups implied by this PR

None filed. The symmetric trap on `badExampleOverride` (empty-string clobber on parsed.badExample) is guarded the same way by the Pipeline 3 call site; no separate ticket needed.

Closes #1580

### CODE REVIEW

File: packages/core/src/compiler-schema.ts [hunk 1]
```diff
@@ -328,9 +351,38 @@ function refineBadExampleRequired(
...
+  if (!data.compilable) return;
+  const engineRequiresGoodExample = data.engine !== 'ast';
+  if (!engineRequiresGoodExample) return;
+  if (typeof data.goodExample === 'string' && data.goodExample.length > 0) return;
+  ctx.addIssue({
+    code: z.ZodIssueCode.custom,
+    message:
+      'goodExample is required (non-empty string) for regex and ast-grep engines (mmnto-ai/totem#1580)',
+    path: ['goodExample'],
+  });
```
[the remark this answers is not in the document]
Comment (satur8d) [reply packages/core/src/compiler-schema.ts:388]: Good catch, fixed in e4f9b361 + caa1df42. Two commits because the same hole lived at both the schema-refine boundary (your flagged line) and the buildCompiledRule gate layer. Shield's round-2 review caught the second one after I pushed the first fix. Both `refineBadExampleRequired`/`refineGoodExampleRequired` and the gate's `hasBadExample`/`hasGoodExample` checks now use `.trim().length > 0`. Fixed the symmetric hole on badExample (pre-existing since #1409) at the same time.

Also added 10 new tests: 6 at the schema layer covering whitespace rejection on both fields and goodExample required-field behavior, 2 at the gate layer pinning whitespace rejection via buildCompiledRule, 2 more pinning the `??` vs empty-string override contract from the prior Shield round.

**Comment by satur8d** [comment]
@coderabbitai: Nit addressed in 16abfd3e. Added three ast-grep negative-path tests to the `goodExample required by engine` suite: missing goodExample on flat `astGrepPattern`, whitespace-only goodExample on flat pattern, missing goodExample on compound `astGrepYamlRule`. 35 schema tests pass.
